time.js

```diff
@@ -1,12 +1,169 @@
 }
 
-module.exports = time;
+/**
+ * A passthrough function that can allow users to call `toZonedDateTime` on
+ * almost anything that can be converted to a time.ZonedDateTime without first
+ * checking to see if it's already a time.ZonedDateTime. 
+ * @returns {time.ZonedDateTime} this
+ */
+time.ZonedDateTime.prototype.toZonedDateTime = function () {
+    return this;
+}
+
+/**
+ * Change the date of the passed in dt to match the date. This preserves
+ * the hours and minutes even in cases where dt and date span a DST changeover.
+ * @param {time.ZonedDateTime} dt date time to move the date on
+ * @param {time.ZonedDateTime} date date time to move dt's date to
+ * @returns {time.ZonedDateTime} dt with date's date
+ */
+const changeDate = function (dt, date) {
+    return dt.withYear(date.year())
+             .withMonth(date.month())
+             .withDayOfMonth(date.dayOfMonth());
+}
+
+/**
+ * Moves the date portion of the date time to today, accounting for DST
+ * @returns {time.ZonedDateTime} this ZDT with today's date
+ */
+time.ZonedDateTime.prototype.toToday = function () {
+    return changeDate(this, time.ZonedDateTime.now());
+}
+
+/**
+ * Moves the date portion of the date time to tomorrow, accounting for DST
+ * @returns {time.ZonedDateTime} this ZDT with tomorrow's date
+ */
+time.ZonedDateTime.prototype.toTomorrow = function () {
+    return changeDate(this, time.ZonedDateTime.now().plusDays(1));
+}
+
+/**
+ * Moves the date portion of the date time to yesterday, accounting for DST
+ * @returns {time.ZonedDateTime} this ZDT with yesterday's date
+ */
+time.ZonedDateTime.prototype.toYesterday = function () {
+    return changeDate(this, time.ZonedDateTime.now().minusDays(1));
+}
+
+/**
+ * Compares this ZDT to see if it falls between start and end times,
+ * accounting for times that span midnight. start and end can be any type
+ * that has a `toZonedDateTime()` method but only the time portion is used.
+ * @param {*} start the starting time, anything that has a toZonedDateTime method
+ * @param {*} end the ending time, anything that has a toZonedDateTime method
+ * @returns {Boolean} true if this is between start and end
+ */
+time.ZonedDateTime.prototype.betweenTimes = function(start, end) {
+    start = start.toZonedDateTime();
+    end = end.toZonedDateTime();
+    if(end.isBefore(start)) {
+        if(this.isAfter(start)) {
+            end = end.toTomorrow();
+        }
+        else if(this.isBefore(start)) {
+            start = start.toYesterday();
+        }
+    }
+    return this.isAfter(start) && this.isBefore(end);
+}
+
+/**
+ * Adds this Duration to now, returning it as a ZonedDateTime
+ * @returns {time.ZonedDateTime} 
+ */
+ time.Duration.prototype.toZonedDateTime = function () {
+    return time.ZonedDateTime.now().plus(this);
+}
+
+/**
+ * @returns {time.ZonedDateTime} value of the number rounded to the nearest int added to now as milliseconds
+ */
+Number.prototype.toZonedDateTime = function() {
```

Comment:
I did so at the suggestion of @jpg. I'm not sure how we would be able to handle the case where one can have something like

    createTimer(value.toZonedDateTime()...

without needing to care whether it's a number of milliseconds, a string of the format "PT3h1m", an openHAB DateTime Item, etc. If we remove those from Number and String I'd say we should probable then revert to having just a static (can you monkey patch a static method) `toZonedDateTime()` or `from` method on `ZonedDateTime`, a stand alone function in `time` (which @jpg doesn't like), or scrap the whole thing.

Reply:
what about having a static function like `time.toZDT(timeObj)` that would convert an object to a ZDT ,  `timeObj` could be:

- Number
- String
- Item
- ZonedDateTime
- some other object

This function could check the object type using `instanceof` and do the same logic  that you have in the Native object monkey patching.   Would that work?

Overriding native types is usually  a bad practice as it leads to maintenance issues in the future, but... i also realize this library is only useful within the OH runtime which we control, so it not completely off the table, i just want us to be really confident  there is a good reason to do it (it has a bad smell) .

Reply:
oh, and i realized we said we didn't not want `time` to be a wasteland of helper functions, but i think having something like this is ok, it's one function that converts all types to our main time format, seems appropriate to me.

Reply:
That's actually how I initially proposed to implement this (and it's how I implemented it in my rules tools library) so it's easy to switch to that approach for sure. The hard part was figuring out the conversions themselves. The rest is just refactoring. ;-)

Should I remove the toZonedDateTime function from Item too so there's just the one function?

What about the toToday, toTomorrow, toYesterday, betweenTimes functions. Should I keep those around as monkey patches or move them to `time` too? 

I'll start working on an update. Gotta fight a PostgreSQL upgrade first though so it might be a couple more days.

Reply:
>Should I remove the toZonedDateTime function from Item too so there's just the one function?

I think so,  it's probably easier to manage and document if it's in one place? 

> What about the toToday, toTomorrow, toYesterday, betweenTimes functions. Should I keep those around as monkey patches or move them to time too?

I thought those seem reasonable to have on a ZonedDateTime object.

Reply:
> I think so, it's probably easier to manage and document if it's in one place?

That would be my argument to do so. 

> I thought those seem reasonable to have on a ZonedDateTime object.

Cool, I'll get a PR submitted as soon as I can. It's mostly just moving stuff around so shouldn't take too long.